plug-ins/prometheus/architecture.md

```diff
@@ -0,0 +1,76 @@
+# Prometheus Plugin Architecture
+
+This document describe the architecture for the Prometheus Plugin (eosio::prometheus_plugin).  The Prometheus Plugin instruments nodes in order to allow
+data to be collected from [Prometheus](https://prometheus.io/).  The requirements for this plugin are described in this [Github issue](https://github.com/eosnetworkfoundation/mandel/issues/67).
+
+## Client Library
+
+A C++ client library exists to facilitate instrumenting a program to export data point to Prometheus called [prometheus-cpp](https://github.com/jupp0r/prometheus-cpp).
+
+Unfortunately, that library is somewhat heavy weight and contains dependencies that are not desirable to bring into the leap codebase, the most problematic being Civetweb:
+
+
+* `civetweb` for [Civetweb](https://github.com/civetweb/civetweb)
+* `com_google_googletest` for [Google Test](https://github.com/google/googletest)
+* `com_github_google_benchmark` for [Google Benchmark](https://github.com/google/benchmark)
+* `com_github_curl` for [curl](https://curl.haxx.se/)
+* `net_zlib_zlib` for [zlib](http://www.zlib.net/)
+
+Instead of making prometheus-cpp a submodule of leap, a new library named prometheus will be created to support exporting of data points from nodeos.
+Code will be ported over from prometheus-cpp, but converted to be more in keeping with the leap coding style (snake case, .hpp & .cpp).  Substantially all of "core" will be ported 
+
+Instead of relying on civetweb to implement the pull functionality, the Prometheus plugin will implement those functions on top of built-in http
+support in nodeos.  
+
+The pull and push functionality will not be pulled into this library, but will be implemented directly in the plugin, due to dependencies on other plugins. 
+
+## Plugin
+
+A new Prometheus plugin will be created.  This plugin will interface with other plugins (chain_plugin, producer_plugin, http_plugin) to supply the data to be exposed.  Instrumentation will be need to be added in the plugin to capture some of these statistics. 
+A call will be added to plugins to get their specific stats (in a non-prometheus dependent fashion).
+
+Data from the original github issue (these data points will require curation/prioritization):
+
+Chain Plugin
+* what is returned from /v1/chain/get_info (head block number, lib)
+* replay status (when starting from snapshot and replaying blocks)
+
```

Comment:
Would be good to have some information about multi-threading approach. Seems like the best approach there would be for the main thread to push data into some kind of sink OR maybe emit a signal for the data that the prometheus plugin will make available. I don't think we want to introduce locking around shared data that the main thread is populating for export by prometheus.